File: python/23.05/61.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if k == 0 or not head:
            return head
        
        first = head
        node = head
        arr = []
        while node:
            arr.append(node)
            last = node
            node = node.next

        for i in range(2, k%len(arr) + 2):
            last.next = first
            first = last
            arr[-i].next = None
            last = arr[-i]
        
        return first

File: typescript/23.08/61.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function rotateRight(head: ListNode | null, k: number): ListNode | null {
    let n = 0
    let first = head
    let last = null
    while(head) {
        if (!head.next) {
            last = head
        }
        head = head.next
        n++
    }

    let cnt = 0
    while (cnt < k % n) {
        last.next = first
        first = last
        let node = first
        let tmpCnt = 0
        while (tmpCnt < n-1) {
            node = node.next
            tmpCnt++
        }
        last = node
        node.next =null
        cnt++
    }

    return first
};
